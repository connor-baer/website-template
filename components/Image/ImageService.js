import { parse, format } from 'url';
import { isNil, omitBy } from 'lodash/fp';

// eslint-disable-next-line import/prefer-default-export
export function getSrc(src, { width: w, height: h, format: fm }) {
  if (!src) {
    return undefined;
  }

  const { query = {}, search, ...rest } = parse(src, true);
  const params = { w, h, fm, url: query.url };
  const truthyParams = omitBy(isNil, params);

  return format({ ...rest, query: truthyParams });
}
