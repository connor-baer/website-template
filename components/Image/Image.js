import React from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import * as ImageService from './ImageService';

const wrapperBaseStyles = () => css`
  display: block;
  position: relative;
  overflow: hidden;
  height: auto;
  width: 100%;
`;

const wrapperAspectRatioStyles = ({ aspectRatio }) =>
  aspectRatio &&
  css`
    height: 0;
    padding-top: ${(1 / aspectRatio) * 100}%;
  `;

const Wrapper = styled('div')(wrapperBaseStyles, wrapperAspectRatioStyles);

const imageBaseStyles = () => css`
  display: block;
  line-height: 0;
  width: 100%;
  height: auto;
  max-width: 100%;
  max-height: 100%;
  color: inherit;
  text-align: center;
`;

const imageAspectRatioStyles = ({ aspectRatio }) =>
  aspectRatio &&
  css`
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    height: 100%;
    object-fit: cover;
  `;

const Img = styled('img')(imageBaseStyles, imageAspectRatioStyles);

export default function Image({ as, src, alt, loading, className, ...props }) {
  const aspectRatio =
    props.width && props.height ? props.width / props.height : false;
  return (
    <Wrapper aspectRatio={aspectRatio} as={as} className={className}>
      <Img
        alt={alt}
        loading={loading}
        src={ImageService.getSrc(src, props)}
        aspectRatio={aspectRatio}
        className={className}
      />
    </Wrapper>
  );
}
