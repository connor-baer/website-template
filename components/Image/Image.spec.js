import React from 'react';

import Image from './Image';

describe('Image', () => {
  function renderImage(fn, props = {}) {
    return fn(<Image {...props} />);
  }

  it('should render with default styles', () => {
    const src = 'https://source.unsplash.com/random';
    const alt = 'A random image from Unsplash';
    const width = 500;
    const { container } = renderImage(render, { src, alt, width });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with fixed aspect ratio styles', () => {
    const src = 'https://source.unsplash.com/random';
    const alt = 'A random image from Unsplash';
    const width = 500;
    const height = 300;
    const { container } = renderImage(render, { src, alt, width, height });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should meet basic accessibility guidelines', async () => {
    const src = 'https://source.unsplash.com/random';
    const alt = 'A random image from Unsplash';
    const html = renderImage(renderToHtml, { src, alt });
    const result = await axe(html);
    expect(result).toHaveNoViolations();
  });
});
