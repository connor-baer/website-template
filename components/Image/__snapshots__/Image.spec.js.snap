// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should render with default styles 1`] = `
.circuit-2 {
  display: block;
  position: relative;
  overflow: hidden;
  height: auto;
  width: 100%;
}

.circuit-0 {
  display: block;
  line-height: 0;
  width: 100%;
  height: auto;
  max-width: 100%;
  max-height: 100%;
  color: inherit;
  text-align: center;
}

<div
  class="circuit-2 circuit-3"
>
  <img
    alt="A random image from Unsplash"
    class="circuit-0 circuit-1"
    src="https://source.unsplash.com/random?w=500"
  />
</div>
`;

exports[`Image should render with fixed aspect ratio styles 1`] = `
.circuit-2 {
  display: block;
  position: relative;
  overflow: hidden;
  height: auto;
  width: 100%;
  height: 0;
  padding-top: 60%;
}

.circuit-0 {
  display: block;
  line-height: 0;
  width: 100%;
  height: auto;
  max-width: 100%;
  max-height: 100%;
  color: inherit;
  text-align: center;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  object-fit: cover;
}

<div
  class="circuit-2 circuit-3"
>
  <img
    alt="A random image from Unsplash"
    class="circuit-0 circuit-1"
    src="https://source.unsplash.com/random?w=500&h=300"
  />
</div>
`;
